File: next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    int n = nums2.size();
    
    unordered_map<int,int>mp;        
    stack<int>st;
    for(int i=n-1;i>=0;i--){
        if(st.empty()){
            st.push(nums2[i]);
            mp[nums2[i]]=-1;
        }
        else{
            while(!st.empty() and st.top()<nums2[i]){
                st.pop();
            }
            if(!st.empty()){
                mp[nums2[i]]=st.top();
            }
            else{
                mp[nums2[i]] = -1;
            }
            st.push(nums2[i]);
        }
    }
    vector<int>res;
    for(int i=0;i<nums1.size();i++){
        res.push_back(mp[nums1[i]]);
    }
    return res;
}
};

File: leetcode/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n = len(nums1)
        stack = [-1]*n

        for i in range(n):
            ind = nums2.index(nums1[i])
            for j in range(ind,len(nums2)):
                if nums2[j] > nums1[i]:
                    stack[i] = nums2[j]
                    break
        return stack 
        